File: backend/src/BuildERP.Infrastructure/Services/InventoryService.cs
using BuildERP.Application.Common.Interfaces;
using BuildERP.Application.Common.Models;
using BuildERP.Application.Features.Inventory;
using BuildERP.Domain.Entities;
using BuildERP.Domain.Enums;
using BuildERP.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace BuildERP.Infrastructure.Services;

public class InventoryService : IInventoryService
{
    private readonly ApplicationDbContext _db;
    public InventoryService(ApplicationDbContext db) => _db = db;

    public async Task<PaginatedList<InventoryItemDto>> GetAllAsync(InventoryQuery query, CancellationToken ct = default)
    {
        var q = _db.InventoryItems.AsQueryable();
        if (!string.IsNullOrWhiteSpace(query.Search))
        {
            var s = query.Search.ToLower();
            q = q.Where(i => i.Name.ToLower().Contains(s) ||
                             (i.Category != null && i.Category.ToLower().Contains(s)));
        }
        if (!string.IsNullOrWhiteSpace(query.Category))
            q = q.Where(i => i.Category == query.Category);
        if (query.LowStock == true)
            q = q.Where(i => i.CurrentStock <= i.LowStockThreshold);

        q = q.OrderBy(i => i.Name);
        var totalCount = await q.CountAsync(ct);
        var items = await q.Skip((query.PageNumber - 1) * query.PageSize).Take(query.PageSize).ToListAsync(ct);
        return new PaginatedList<InventoryItemDto>(items.Select(MapToDto).ToList(), totalCount, query.PageNumber, query.PageSize);
    }

    public async Task<InventoryItemDto> GetByIdAsync(Guid id, CancellationToken ct = default)
    {
        var entity = await _db.InventoryItems.FindAsync(new object[] { id }, ct)
            ?? throw new KeyNotFoundException($"Item {id} not found.");
        return MapToDto(entity);
    }

    public async Task<InventoryItemDto> CreateAsync(CreateInventoryItemRequest request, CancellationToken ct = default)
    {
        var entity = new InventoryItem
        {
            Name = request.Name, Category = request.Category, Unit = request.Unit,
            CurrentStock = request.CurrentStock, LowStockThreshold = request.LowStockThreshold,
            UnitPrice = request.UnitPrice, SupplierName = request.SupplierName,
            Location = request.Location, Notes = request.Notes
        };
        _db.InventoryItems.Add(entity);
        await _db.SaveChangesAsync(ct);
        return MapToDto(entity);
    }

    public async Task<InventoryItemDto> UpdateAsync(Guid id, UpdateInventoryItemRequest request, CancellationToken ct = default)
    {
        var entity = await _db.InventoryItems.FindAsync(new object[] { id }, ct)
            ?? throw new KeyNotFoundException($"Item {id} not found.");
        entity.Name = request.Name; entity.Category = request.Category; entity.Unit = request.Unit;
        entity.LowStockThreshold = request.LowStockThreshold; entity.UnitPrice = request.UnitPrice;
        entity.SupplierName = request.SupplierName; entity.Location = request.Location; entity.Notes = request.Notes;
        await _db.SaveChangesAsync(ct);
        return MapToDto(entity);
    }

    public async Task DeleteAsync(Guid id, CancellationToken ct = default)
    {
        var entity = await _db.InventoryItems.FindAsync(new object[] { id }, ct)
            ?? throw new KeyNotFoundException($"Item {id} not found.");
        _db.InventoryItems.Remove(entity);
        await _db.SaveChangesAsync(ct);
    }

    public async Task<List<StockTransactionDto>> GetTransactionsAsync(Guid itemId, CancellationToken ct = default)
    {
        var txns = await _db.StockTransactions
            .Include(t => t.InventoryItem)
            .Where(t => t.InventoryItemId == itemId)
            .OrderByDescending(t => t.Date)
            .ToListAsync(ct);
        return txns.Select(MapTxnToDto).ToList();
    }

    public async Task<StockTransactionDto> AddTransactionAsync(Guid itemId, AddStockTransactionRequest request, CancellationToken ct = default)
    {
        var item = await _db.InventoryItems.FindAsync(new object[] { itemId }, ct)
            ?? throw new KeyNotFoundException($"Item {itemId} not found.");

        var txn = new StockTransaction
        {
            InventoryItemId = itemId,
            Type = request.Type,
            Quantity = request.Quantity,
            UnitPrice = request.UnitPrice,
            Date = request.Date,
            Reference = request.Reference,
            ProjectName = request.ProjectName,
            Notes = request.Notes
        };

        if (request.Type == StockTransactionType.StockIn)
            item.CurrentStock += request.Quantity;
        else
            item.CurrentStock -= request.Quantity;

        _db.StockTransactions.Add(txn);
        await _db.SaveChangesAsync(ct);
        txn.InventoryItem = item;
        return MapTxnToDto(txn);
    }

    private static InventoryItemDto MapToDto(InventoryItem i) => new()
    {
        Id = i.Id, Name = i.Name, Category = i.Category, Unit = i.Unit,
        CurrentStock = i.CurrentStock, LowStockThreshold = i.LowStockThreshold,
        UnitPrice = i.UnitPrice, SupplierName = i.SupplierName, Location = i.Location,
        IsLowStock = i.CurrentStock <= i.LowStockThreshold,
        Notes = i.Notes, CreatedAt = i.CreatedAt
    };

    private static StockTransactionDto MapTxnToDto(StockTransaction t) => new()
    {
        Id = t.Id, InventoryItemId = t.InventoryItemId,
        ItemName = t.InventoryItem?.Name ?? string.Empty,
        Type = t.Type, TypeDisplay = t.Type.ToString(),
        Quantity = t.Quantity, UnitPrice = t.UnitPrice, Date = t.Date,
        Reference = t.Reference, ProjectName = t.ProjectName, Notes = t.Notes, CreatedAt = t.CreatedAt
    };
}

File: backend/src/inventory/inventory.service.ts
import { Injectable, NotFoundException, BadRequestException, ConflictException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import {
  CreateInventoryItemDto,
  UpdateInventoryItemDto,
  AddStockTransactionDto,
  InventoryQueryDto,
} from './dto/inventory.dto';
import { StockTransactionType } from '@prisma/client';
import { generateCode } from '../common/utils/generate-code';

@Injectable()
export class InventoryService {
  constructor(private readonly prisma: PrismaService) {}

  async findAll(query: InventoryQueryDto) {
    const page = Math.max(1, Math.floor(query.pageNumber ?? 1));
    const limit = Math.min(100, Math.max(1, Math.floor(query.pageSize ?? 10)));
    const skip = (page - 1) * limit;

    const where: any = {};

    if (query.search) {
      where.OR = [
        { code: { contains: query.search } },
        { name: { contains: query.search } },
        { category: { contains: query.search } },
        { supplierName: { contains: query.search } },
      ];
    }

    if (query.category) {
      where.category = { contains: query.category };
    }

    // lowStock filter: currentStock <= lowStockThreshold
    // Use $queryRaw for column-to-column comparison which Prisma's findMany does not support.
    if (query.lowStock === true) {
      const items = await this.prisma.$queryRaw<any[]>`
        SELECT * FROM inventory_items
        WHERE current_stock <= low_stock_threshold
        ORDER BY current_stock ASC
        LIMIT ${limit} OFFSET ${skip}
      `;
      const totalCount = await this.prisma.$queryRaw<any[]>`
        SELECT COUNT(*) as count FROM inventory_items
        WHERE current_stock <= low_stock_threshold
      `;

      const total = Number(totalCount[0].count);
      const totalPages = Math.ceil(total / limit);
      const pageNumber = page;
      const pageSize = limit;

      return {
        items: items.map((item) => this.mapInventoryItem({
          id: item.id,
          code: item.code,
          name: item.name,
          category: item.category,
          unit: item.unit,
          currentStock: item.current_stock,
          lowStockThreshold: item.low_stock_threshold,
          unitPrice: item.unit_price,
          supplierName: item.supplier_name,
          location: item.location,
          notes: item.notes,
          createdAt: new Date(item.created_at),
        })),
        totalCount: total,
        pageNumber,
        pageSize,
        totalPages,
        hasPreviousPage: pageNumber > 1,
        hasNextPage: pageNumber < totalPages,
      };
    }

    const [inventoryItems, total] = await Promise.all([
      this.prisma.inventoryItem.findMany({
        where,
        skip,
        take: limit,
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.inventoryItem.count({ where }),
    ]);

    const totalPages = Math.ceil(total / limit);
    const pageNumber = page;
    const pageSize = limit;

    return {
      items: inventoryItems.map((item) => this.mapInventoryItem(item)),
      totalCount: total,
      pageNumber,
      pageSize,
      totalPages,
      hasPreviousPage: pageNumber > 1,
      hasNextPage: pageNumber < totalPages,
    };
  }

  async findById(id: string) {
    const item = await this.prisma.inventoryItem.findUnique({
      where: { id },
      include: {
        stockTransactions: {
          orderBy: { date: 'desc' },
        },
      },
    });

    if (!item) {
      throw new NotFoundException(`Inventory item with id ${id} not found`);
    }

    return {
      ...this.mapInventoryItem(item),
      stockTransactions: item.stockTransactions.map((t) =>
        this.mapStockTransaction(t),
      ),
    };
  }

  async getNextCode(): Promise<string> {
    return generateCode(this.prisma, 'inventoryItem');
  }

  async create(dto: CreateInventoryItemDto) {
    let code: string;
    if (dto.code) {
      const existing = await this.prisma.inventoryItem.findUnique({ where: { code: dto.code } });
      if (existing) throw new ConflictException('Code already in use');
      code = dto.code;
    } else {
      code = await generateCode(this.prisma, 'inventoryItem');
    }
    const item = await this.prisma.inventoryItem.create({
      data: {
        code,
        name: dto.name,
        ...(dto.category !== undefined && { category: dto.category }),
        ...(dto.unit !== undefined && { unit: dto.unit }),
        currentStock: dto.currentStock ?? 0,
        lowStockThreshold: dto.lowStockThreshold ?? 0,
        unitPrice: dto.unitPrice ?? 0,
        ...(dto.supplierName !== undefined && { supplierName: dto.supplierName }),
        ...(dto.location !== undefined && { location: dto.location }),
        ...(dto.notes !== undefined && { notes: dto.notes }),
      },
    });

    return this.mapInventoryItem(item);
  }

  async update(id: string, dto: UpdateInventoryItemDto) {
    await this.findById(id);

    if (dto.code !== undefined) {
      const conflict = await this.prisma.inventoryItem.findFirst({ where: { code: dto.code, NOT: { id } } });
      if (conflict) throw new ConflictException('Code already in use');
    }

    const item = await this.prisma.inventoryItem.update({
      where: { id },
      data: {
        ...(dto.code !== undefined && { code: dto.code }),
        ...(dto.name !== undefined && { name: dto.name }),
        ...(dto.category !== undefined && { category: dto.category }),
        ...(dto.unit !== undefined && { unit: dto.unit }),
        ...(dto.currentStock !== undefined && { currentStock: dto.currentStock }),
        ...(dto.lowStockThreshold !== undefined && {
          lowStockThreshold: dto.lowStockThreshold,
        }),
        ...(dto.unitPrice !== undefined && { unitPrice: dto.unitPrice }),
        ...(dto.supplierName !== undefined && { supplierName: dto.supplierName }),
        ...(dto.location !== undefined && { location: dto.location }),
        ...(dto.notes !== undefined && { notes: dto.notes }),
      },
    });

    return this.mapInventoryItem(item);
  }

  async remove(id: string) {
    await this.findById(id);

    await this.prisma.inventoryItem.delete({ where: { id } });

    return { message: 'Inventory item deleted successfully' };
  }

  async getTransactions(id: string) {
    // Verify item exists (findById throws NotFoundException if not found)
    const item = await this.prisma.inventoryItem.findUnique({ where: { id } });

    if (!item) {
      throw new NotFoundException(`Inventory item with id ${id} not found`);
    }

    const transactions = await this.prisma.stockTransaction.findMany({
      where: { inventoryItemId: id },
      orderBy: { date: 'desc' },
    });

    return transactions.map((t) => this.mapStockTransaction(t));
  }

  async addTransaction(
    id: string,
    dto: AddStockTransactionDto,
    userId: string,
  ) {
    const item = await this.prisma.inventoryItem.findUnique({ where: { id } });

    if (!item) {
      throw new NotFoundException(`Inventory item with id ${id} not found`);
    }

    const currentStock = Number(item.currentStock);
    const quantity = Number(dto.quantity);
    let newStock: number;

    if (dto.type === StockTransactionType.In) {
      newStock = currentStock + quantity;
    } else if (dto.type === StockTransactionType.Out) {
      newStock = currentStock - quantity;
      if (newStock < 0) {
        throw new BadRequestException('Insufficient stock');
      }
    } else {
      // Adjustment: set stock directly to quantity
      newStock = quantity;
    }

    await this.prisma.stockTransaction.create({
      data: {
        inventoryItemId: id,
        type: dto.type,
        quantity: dto.quantity,
        ...(dto.unitPrice !== undefined && { unitPrice: dto.unitPrice }),
        date: new Date(dto.date),
        ...(dto.reference !== undefined && { reference: dto.reference }),
        ...(dto.projectName !== undefined && { projectName: dto.projectName }),
        ...(dto.notes !== undefined && { notes: dto.notes }),
        ...(userId && { createdById: userId }),
      },
    });

    const updatedItem = await this.prisma.inventoryItem.update({
      where: { id },
      data: { currentStock: newStock },
    });

    return this.mapInventoryItem(updatedItem);
  }

  private mapInventoryItem(item: {
    id: string;
    code?: string | null;
    name: string;
    category: string | null;
    unit: string | null;
    currentStock: any;
    lowStockThreshold: any;
    unitPrice: any;
    supplierName: string | null;
    location: string | null;
    notes: string | null;
    createdAt: Date;
  }) {
    const currentStock = Number(item.currentStock);
    const lowStockThreshold = Number(item.lowStockThreshold);

    return {
      id: item.id,
      code: item.code ?? null,
      name: item.name,
      category: item.category,
      unit: item.unit,
      currentStock,
      lowStockThreshold,
      unitPrice: Number(item.unitPrice),
      supplierName: item.supplierName,
      location: item.location,
      isLowStock: currentStock <= lowStockThreshold,
      notes: item.notes,
      createdAt: item.createdAt,
    };
  }

  private mapStockTransaction(tx: {
    id: string;
    inventoryItemId: string;
    type: any;
    quantity: any;
    unitPrice: any;
    date: Date;
    reference: string | null;
    projectName: string | null;
    notes: string | null;
    createdAt: Date;
    createdById: string | null;
  }) {
    return {
      id: tx.id,
      inventoryItemId: tx.inventoryItemId,
      type: tx.type,
      quantity: Number(tx.quantity),
      unitPrice: tx.unitPrice !== null ? Number(tx.unitPrice) : null,
      date: tx.date,
      reference: tx.reference,
      projectName: tx.projectName,
      notes: tx.notes,
      createdById: tx.createdById,
      createdAt: tx.createdAt,
    };
  }
}
